crates/turbopack-core/src/chunk/mod.rs

```diff
@@ -135,41 +150,79 @@ impl ChunkGroupVc {
+        Self::from_chunk(
+            chunking_context,
+            asset.as_chunk(chunking_context, availability_info),
+        )
     }
 
-    /// Creates a chunk group from an chunk as entrypoint
+    /// Creates a chunk group from a chunk as entrypoint
     #[turbo_tasks::function]
-    pub fn from_chunk(chunk: ChunkVc) -> Self {
-        Self::cell(ChunkGroup { entry: chunk })
+    pub fn from_chunk(chunking_context: ChunkingContextVc, entry: ChunkVc) -> Self {
+        Self::cell(ChunkGroup {
+            chunking_context,
+            entry,
+            evaluated_entries: EvaluatedEntriesVc::empty(),
+        })
     }
 
-    /// Returns the entry chunk of this chunk group.
+    /// Creates a chunk group from a chunk as entrypoint, with the given
+    /// evaluated entries to be appended.
+    ///
+    /// `main_entry` will always be evaluated after all entries in
+    /// `other_entries` are evaluated.
     #[turbo_tasks::function]
-    pub async fn entry(self) -> Result<ChunkVc> {
-        Ok(self.await?.entry)
+    pub fn evaluated(
+        chunking_context: ChunkingContextVc,
+        main_entry: EvaluatedEntryVc,
+        other_entries: EvaluatedEntriesVc,
+    ) -> Self {
+        Self::cell(ChunkGroup {
+            chunking_context,
+            entry: main_entry.as_root_chunk(chunking_context),
+            // The main entry should always be *appended* to other entries, in order to ensure
+            // it's only evaluated once all other entries are evaluated.
+            evaluated_entries: other_entries.with_entry(main_entry),
+        })
     }
 
-    /// Returns the chunk list path for this chunk group.
+    /// Returns the entry chunk of this chunk group.
     #[turbo_tasks::function]
-    pub async fn chunk_list_path(self) -> Result<FileSystemPathVc> {
-        let this = self.await?;
-        Ok(this
-            .entry
-            .chunking_context()
-            .chunk_list_path(this.entry.ident().path()))
+    pub async fn entry(self) -> Result<ChunkVc> {
+        Ok(self.await?.entry)
     }
 
     /// Lists all chunks that are in this chunk group.
     /// These chunks need to be loaded to fulfill that chunk group.
     /// All chunks should be loaded in parallel.
     #[turbo_tasks::function]
-    pub async fn chunks(self) -> Result<ChunksVc> {
+    pub async fn chunks(self) -> Result<AssetsVc> {
+        let this = self.await?;
+        let evaluated_entries = this.evaluated_entries.await?;
+
+        let mut entry_chunks: HashSet<_> = evaluated_entries
+            .iter()
+            .map({
+                let chunking_context = this.chunking_context;
+                move |evaluated_entry| async move {
+                    Ok(evaluated_entry
+                        .as_root_chunk(chunking_context)
+                        .resolve()
+                        .await?)
+                }
+            })
+            .try_join()
+            .await?
+            .into_iter()
+            .collect();
```

Comment:
This is a big difference: before, we were putting evaluated entries inside of the main entries of the "evaluated chunk". Now, they are considered separate chunks to begin with, and it's up to the optimizer to merge them.